src/python/pants/rules/core/list_roots.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Set
+
+from pants.engine.console import Console
+from pants.engine.fs import PathGlobs, Snapshot
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.rules import console_rule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.source.source_root import AllExistingSourceRoots, SourceRoot, SourceRootConfig
+
+
+class Roots(LineOriented, Goal):
+  """List the repo's registered source roots."""
+  name = 'roots'
+
+
+@rule(AllExistingSourceRoots, [SourceRootConfig])
+def all_roots(source_root_config):
+
+  source_roots = source_root_config.get_source_roots()
+
+  all_paths: Set[str] = set()
+  for path in source_roots.traverse():
+    if path.startswith("^/"):
+      all_paths |= {f"{path[2:]}/"}
```

Comment:
I think `all_paths.add()` is preferable instead of constructing a throwaway singleton set.